### PR TITLE
reference certificate by name in ALB spec

### PR DESCRIPTION
**Pull request by robfletcher**
Instead of requiring users to specify the certificate ARN, we can resolve that from Clouddriver so users can reference the cert by name. This PR adds a new ALB spec version.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -62,6 +62,7 @@ subprojects {
 
       "testImplementation"("org.junit.platform:junit-platform-runner")
       "testImplementation"("org.junit.jupiter:junit-jupiter-api")
+      "testImplementation"("org.junit.jupiter:junit-jupiter-params")
```
Comment (robfletcher) [inline build.gradle.kts:65]: Just applying this to all sub-projects instead of having to add it to each individually

File: keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/CloudDriverCache.kt [hunk 2]
```diff
@@ -32,6 +33,8 @@ interface CloudDriverCache {
   fun credentialBy(name: String): Credential
   fun defaultKeyPairForAccount(account: String) =
     credentialBy(account).attributes["defaultKeyPair"] as String
+  fun certificateByName(name: String): Certificate
+  fun certificateByArn(arn: String): Certificate
```
Comment (robfletcher) [inline keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/CloudDriverCache.kt:37]: New CloudDriverCache methods

File: keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/CloudDriverService.kt [hunk 3]
```diff
@@ -196,4 +197,7 @@ interface CloudDriverService {
     @Query("entityId") entityId: String,
     @Header("X-SPINNAKER-USER") user: String = DEFAULT_SERVICE_ACCOUNT
   ): List<EntityTags>
+
+  @GET("/certificates/aws")
+  suspend fun getCertificates() : List<Certificate>
```
Comment (robfletcher) [inline keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/CloudDriverService.kt:202]: River endpoint we can pull certs from. This returns _all_ certificates at once

File: keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/MemoryCloudDriverCache.kt [hunk 4]
```diff
@@ -158,6 +159,36 @@ class MemoryCloudDriverCache(
...
+          cloudDriver
+            .getCertificates()
+            .find { it.serverCertificateName == name }
+            ?.also {
+              certificatesByArn.put(it.arn, completedFuture(it))
+            }
+        }
+          .handleNotFound()
+          ?: notFound("Certificate with name $name not found")
+      }
```
Comment (robfletcher) [inline keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/MemoryCloudDriverCache.kt:175]: New caches for certs. This pattern is a bit screwy for endpoints like this that return all the data in a single payload. We're doing the same thing for accounts, subnets, AZs, etc. I'm going to try to pull together another PR with a better approach.

File: keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/model/Certificate.kt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+package com.netflix.spinnaker.keel.clouddriver.model
+
+data class Certificate(val serverCertificateName: String, val arn: String)
```
Comment (robfletcher) [inline keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/model/Certificate.kt:3]: Certificate as returned by the River endpoint

File: keel-ec2-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ApplicationLoadBalancerSpec.kt [hunk 6]
```diff
@@ -39,10 +36,18 @@ data class ApplicationLoadBalancerSpec(
...
-  )
+  ) {
+    init {
+      if (protocol == "HTTPS") {
+        requireNotNull(certificate) {
+          "HTTPS listeners must specify a certificate"
+        }
+      }
+    }
+  }
```
Comment (robfletcher) [inline keel-ec2-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ApplicationLoadBalancerSpec.kt:50]: Change to the listener spec 1) field is the certificate name rather than ARN, 2) certificate is required if using HTTPS

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/migrators/ApplicationLoadBalancerV1_1ToV1_2Migrator.kt [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      moniker = spec.moniker,
+      locations = spec.locations,
+      internal = spec.internal,
+      dependencies = spec.dependencies,
+      idleTimeout = spec.idleTimeout,
+      listeners = spec.listeners.mapTo(mutableSetOf()) {
+        ApplicationLoadBalancerSpec.Listener(
+          port = it.port,
+          protocol = it.protocol,
+          certificate = it.certificateArn?.let { cloudDriverCache.certificateByArn(it).serverCertificateName },
```
Comment (robfletcher) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/migrators/ApplicationLoadBalancerV1_1ToV1_2Migrator.kt:34]: Migrator uses the new cache to look up the certificate name

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ApplicationLoadBalancerHandler.kt [hunk 8]
```diff
@@ -227,12 +234,16 @@ class ApplicationLoadBalancerHandler(
...
+                    "internal",
+                    ignoreCase = true
+                  ),
                   listeners = lb.listeners.map { l ->
                     ApplicationLoadBalancerSpec.Listener(
                       port = l.port,
                       protocol = l.protocol,
-                      certificateArn = l.certificates?.firstOrNull()?.certificateArn,
+                      certificate = l.certificates?.firstOrNull()
+                        ?.let { cloudDriverCache.certificateByArn(it.certificateArn).serverCertificateName },
```
Comment (robfletcher) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ApplicationLoadBalancerHandler.kt:246]: Use the cache to translate the ARN included in the River ALB response to the cert name

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ApplicationLoadBalancerHandler.kt [hunk 9]
```diff
@@ -301,11 +312,15 @@ class ApplicationLoadBalancerHandler(
...
-              }
+              it.certificate?.let { certificateName ->
+                this + mapOf(
+                  "certificates" to listOf(
+                    mapOf(
+                      "certificateArn" to cloudDriverCache.certificateByName(certificateName).arn
+                    )
+                  )
+                )
+              } ?: this
```
Comment (robfletcher) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ApplicationLoadBalancerHandler.kt:323]: In the Orca payload we need to specify the certificate ARN

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/jackson/KeelEc2ApiModule.kt [hunk 10]
```diff
@@ -72,7 +73,8 @@ internal object KeelEc2ApiModule : SimpleModule("Keel EC2 API") {
   override fun setupModule(context: SetupContext) {
     with(context) {
       setMixInAnnotations<ApplicationLoadBalancerSpec, ApplicationLoadBalancerSpecMixin>()
-      // same annotations are required for this legacy model, so it can reuse the same mixin
+      // same annotations are required for this legacy models, so they can reuse the same mixin
```
Comment (luispollo) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/jackson/KeelEc2ApiModule.kt:76]:
```suggestion
      // same annotations are required for these legacy models, so they can reuse the same mixin
```

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/migrators/ApplicationLoadBalancerV1_1ToV1_2Migrator.kt [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+          healthCheckProtocol = it.healthCheckProtocol,
+          healthCheckHttpCode = it.healthCheckHttpCode,
+          healthCheckPath = it.healthCheckPath,
+          healthCheckInterval = it.healthCheckInterval,
+          healthyThresholdCount = it.healthyThresholdCount,
+          unhealthyThresholdCount = it.unhealthyThresholdCount,
+          attributes = it.attributes
+        )
+      },
+      overrides = spec.overrides
```
Comment (luispollo) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/migrators/ApplicationLoadBalancerV1_1ToV1_2Migrator.kt:57]: Can't the overrides have listener specs too?

**Review by luispollo — COMMENTED** [review]
PR looks good to me, but I had a question about the certificate names: IIRC, the name is just a tag that you apply to the certificate, and like all tags, it's optional, meaning you can have a certificate without a name. Do you know how these are provisioned at Netflix and whether we enforce usage of that tag?

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/migrators/ApplicationLoadBalancerV1_1ToV1_2Migrator.kt (hunk 11 above)
Comment (robfletcher) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/migrators/ApplicationLoadBalancerV1_1ToV1_2Migrator.kt:57]: oh yeah, that would make sense

**Comment by robfletcher** [comment]
> PR looks good to me, but I had a question about the certificate names: IIRC, the name is just a tag that you apply to the certificate, and like all tags, it's optional, meaning you can have a certificate without a name. Do you know how these are provisioned at Netflix and whether we enforce usage of that tag?

Based on how Deck works when assigning certs I believe we require it

**Review by luispollo — APPROVED** [review]
LGTM!
